runtime/doc/develop.txt

```diff
@@ -166,77 +166,282 @@ stick to these rules, to keep the sources readable and maintainable.
-   solution.  Send an email to the |vim-dev| maillist with an explanation and
-   include the diff. Or create a pull request on github.
-
-
 C COMPILER				*style-compiler* *ANSI-C* *C89* *C99*
 
 The minimal C compiler version supported is C89, also known as ANSI C.
 Later standards, such as C99, are not widely supported, or at least not 100%
 supported.  Therefore we use only some of the C99 features and explicitly
 disallow some (this will gradually be adjusted over time).
 
-Please don't make changes everywhere to use the C99 features, it causes merge
-problems for existing patches.  Only use them for new and changed code.
+Features not to be used ~
 
-Comments ~
+These C99 features are not to be used, because not enough compilers support
+them:
+- Variable length arrays (even in C11 this is an optional feature).
+- C99 _Bool and _Complex types.
+- "inline" (it's hardly ever needed, let the optimizer do its work)
+- flexible array members: Not supported by HP-UX C compiler (John Marriott)
 
-Traditionally Vim uses /* comments */.  We intend to keep it that way
-for file and function headers and larger blocks of code, E.g.:
+COMMENTS					       *style-comments*
+
+Comments are good, but there is also a danger of over-commenting. NEVER try to
+explain HOW your code works in a comment. Write the code so that the working
+is obvious.
+
+Generally, you want your comments to tell WHAT your code does and WHY it does
+it, instead of HOW it is done.
+
+Try to avoid putting multiline comments inside a function body: if the
+function is so complex that you need to separately comment parts of it, you
+should probably rethink the structure of the function.
+
+Avoid using a comment to explain what a variable is, instead of naming the
+variable sensibly.
+
+For multiline comments use:
+>
+    /*
+     * comment
+     */
+<
+For inline comments use:
+>
+    int var; // comment
+<
+
+
+INDENTATION					      *style-indentation*
+
+We use 4 space to indent the code. If you are using vim, to edit the source,
+you don't need to do anything due to the |modeline|.
+
+For other editors an editorconfig is provided at the root of the repo.
+
+
+DECLARATIONS	                                     *style-declarations*
+
+Declare for loop variable in the guard (see also |style-loops|):
+Ok:
+>
+    for (int i = 0; i < len; ++i)
+<
+Wrong:
+>
+    int i;
+    for (i = 0; i < len; ++i)
+<
+Always declare a variable with a default value:
+Ok:
+>
+    int n = 0;
+    int *ptr = NULL;
+<
+Wrong:
+>
+    int n;
+    int *ptr;
+<
+
+
+LOOPS						            *style-loops*
+
+Use while loops when:
+- You need to iterate over a data structure that needs to move some sort of
+  pointer, such as a linked list.
+- You need to iterate indefinatly (aka while (TRUE)).
+
+Use for loops when:
+- You need to iterate n time, and you don't need to modify the variable that
+  index
+- There is a for each macro ready to use for your purpose
```

Comment:
The name of that macro can be wrapped in \`\`.

Reply:
Where?

Reply:
You have lost me here.  If there is a suitable `FOR*` macro  
defined, you would probably try using that, otherwise what  
advice do you give: a) define your own `FOR*` macro and use  
that, b) use a `for` statement in the absence of a suitable  
`FOR*` macro?  As it stands, the sentence rather implies:  
‘prefer `for` statements to `FOR*` macros’.  Should it be  
re-worded as:

> There is NO for-each macro ready to use for your purpose

?

Overall, I think, what needs to be spelled out about loops  
is only that:

1) there is a trend (?) to use `while (TRUE) ;` (and  
`do { ; } while (TRUE);`) (17 instances) instead of `for (;;) ;`  
(282 instances) for infinite iteration.

2) look for `FOR*` macros defined in `macros.h` or elsewhere.

Reply:
What I'm trying to say is that you should prefer while loops for infinite iterations and use for loops only when you use the variable in the guard or if there is a `FOR*` macro.

This is not something we are currently doing and it is something I have an issue with.

I've seen a lot of this:
```c
for (; cond; incr)
```
Or this
```c
for (;;)
```
And I think is those circumstances it is better to use a while loop.

I've seen a lot of time in the codebase this obsession with using a for loop no matter what. Sometimes even at the cost of readability and charity of the operation. So I think it is a good idea to write when is appropriate to use a type of loop over the other.

Reply:
I've removed this section that it is overly pedantic